lib/server.js

```diff
@@ -263,27 +263,12 @@ class Server extends KarmaEventEmitter {
       })
 
       socket.on('register', (info) => {
-        let newBrowser = info.id ? (capturedBrowsers.getById(info.id) || singleRunBrowsers.getById(info.id)) : null
-
-        if (newBrowser) {
-          // By default if a browser disconnects while still executing, we assume that the test
-          // execution still continues because just the socket connection has been terminated. Now
-          // since we know whether this is just a socket reconnect or full client reconnect, we
-          // need to update the browser state accordingly. This is necessary because in case a
-          // browser crashed and has been restarted, we need to start with a fresh execution.
-          if (!info.isSocketReconnect) {
-            newBrowser.setState(Browser.STATE_DISCONNECTED)
-          }
-
-          newBrowser.reconnect(socket)
+        const knownBrowser = info.id ? (capturedBrowsers.getById(info.id) || singleRunBrowsers.getById(info.id)) : null
 
-          // Since not every reconnected browser is able to continue with its previous execution,
-          // we need to start a new execution in case a browser has restarted and is now idling.
-          if (newBrowser.state === Browser.STATE_CONNECTED && config.singleRun) {
-            newBrowser.execute(config.client)
-          }
+        if (knownBrowser) {
+          knownBrowser.reconnect(socket, info.isSocketReconnect)
         } else {
-          newBrowser = this._injector.createChild([{
+          const newBrowser = this._injector.createChild([{
```

Comment:
Question: why do we register the new browser in the DI module at this point? When and where will it be required/injected?

Reply:
This obscure code should be cleaned up.  The server can request a new Browser from eg Browser module or BrowserCollection.  Next time.